src/app/website/posts/zkp.markdown

```diff
@@ -0,0 +1,74 @@
+date: 2019-05-13
+author: Vanishree Rao
+---
+
+If you are hearing about zero-knowledge proofs -- a fascinating cryptography tool, for the first time, you are likely to say "Hah! That sounds impossible." Read on to get an intuitive understanding of what they are. But first, some background. Zero-knowledge proofs (or ZKPs)  were [invented](http://people.csail.mit.edu/silvio/Selected%20Scientific%20Papers/Proof%20Systems/The_Knowledge_Complexity_Of_Interactive_Proof_Systems.pdf) by Shafi Goldwasser, Silvio Micali, and Charles Rackoff in 1985. Ever since, ZKPs have been one of the most active areas of research in Cryptography. Moreover, recently, they are enjoying significant impact on real-world applications, specifically on blockchain technologies. [Zcash](https://z.cash/technology/), a pioneering blockchain project, employed ZKPs to achieve anonymity in financial transactions. At O(1)Labs, we are building CODA, the first succinct blockchain, using ZKPs. No matter how many transactions are recorded on the blockchain, the blockchain remains at most the size of a few tweets. 
+
+
+
+## What are Zero-knowledge Proofs?
+
+ZKPs are a powerful tool that enable you do the following. Let’s say you have a secret. Using ZKPs, you can prove that you have the secret without even revealing the it!  
+
+Now let’s look at a puzzle and see how you could convince anybody that you know a solution without revealing it. 
+
+
+### A Puzzle called 3-Coloring. 
+
+The 3-coloring puzzle can be described as follows. You are given a graph of nodes and edges (like in the figure below). The task is to find a "3-coloring" to the graph, which is a coloring of the nodes with three different colors in such a way that no two connected nodes have the same color. 
+
+![An uncolored graph.](/static/blog/zkp/uncolored-graph.png)
+
+![A colored graph.](/static/blog/zkp/colored-graph.png)
+
+### A Zero-knowledge Proof for the 3-coloring Puzzle. 
+
+We will construct a ZKP protocol for the 3-coloring puzzle. Before that, let’s quickly recall the two properties we are looking for in the protocol. 
+
+A ZKP protocol between you and someone -- call her  Verifier, must satisfy the following properties: 
+\begin{itemize}
+\item if you are cheating (i.e., if you do not know a 3-coloring), then Verifier should be able to catch you -- this property is called  \emph{soundness}
+\item Verifier should not learn anything about the 3-coloring -- this property is called  \emph{zero knowledgeness}
+
+Now, let's try to construct a ZKP protocol. Imagine the graph drawn on the floor in a closed space. Per the 3-coloring you know, you would place the corresponding colored balls on the nodes. Then you would completely cover the balls with inverted opaque bowls. 
+Then, Verifier comes and points at an edge of her choice. You would lift up the two bowls on either side of the edge. Verifier verifies that the balls revealed are of different colors. If they are not, you are caught cheating. 
```

Comment:
This would be nice to show with two illustrations -- maybe we can ask our design intern to make them?

Reply:
Great idea. Let me check with her and Harold.

Reply:
done.